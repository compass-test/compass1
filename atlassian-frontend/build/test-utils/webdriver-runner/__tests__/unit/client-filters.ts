import {
  getBrowserTestCaseClients,
  getBetaBrowserTestCaseClients,
  getMobileTestCaseClients,
} from '../../lib/clients/client-filters';
import {
  setBrowserStackDesktopClients,
  setBetaBrowserStackDesktopClients,
  setBrowserStackMobileClients,
} from '../../lib/clients/setup-clients';

function flattenClientToBrowserName(clients: any[]) {
  return clients.map(client => {
    const { capabilities } = client.options;
    return capabilities.browserName.toLowerCase();
  });
}

function flattenClientToBrowserVersion(clients: any[]) {
  return clients.map(client => {
    const { browserName, browserVersion } = client.options.capabilities;
    return `${browserName}${
      browserVersion.includes('beta') ? ' beta' : ' stable'
    }`;
  });
}

function flattenClientToDeviceName(clients: any[]) {
  return clients.map(client => {
    const { os, os_version, device } = client.options.capabilities;
    return `${os} ${os_version} ${device}`;
  });
}

describe('BrowserStack', () => {
  // Desktop Browsers
  describe('getBrowserTestCaseClients', () => {
    const clients = setBrowserStackDesktopClients();
    it('should include only desktop browsers', () => {
      const execClients = getBrowserTestCaseClients(clients);
      const browsers = flattenClientToBrowserName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it('should filter out browser(s) when skipped via test case options', () => {
      const execClients = getBrowserTestCaseClients(clients, {
        skip: ['chrome'],
      });
      const browsers = flattenClientToBrowserName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it('should filter out all browsers when wildcard is passed via test case options', () => {
      const execClients = getBrowserTestCaseClients(clients, {
        skip: ['*'],
      });
      const browsers = flattenClientToBrowserName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it('should ignore invalid skipped browsers via test case options', () => {
      // @ts-ignore deliberately testing invalid browser name since this package is hybrid JS/TS.
      const execClients = getBrowserTestCaseClients(clients, { skip: ['foo'] });
      const browsers = flattenClientToBrowserName(execClients);
      expect(browsers).toMatchSnapshot();
    });
  });

  // Beta Desktop Browsers
  describe('getBetaBrowserTestCaseClients', () => {
    let originalBetaBrowsers = '';
    let clients: any[];

    beforeAll(() => {
      originalBetaBrowsers = process.env.BETA_BROWSERS || '';
      process.env.BETA_BROWSERS = `[{"os":"Windows","os_version":"10","browserName":"chrome","browser_version":"85.0 beta","RESOLUTION":"1920x1080"}]`;
      clients = setBetaBrowserStackDesktopClients();
    });

    it('should include only beta desktop browsers', () => {
      const execClients = getBetaBrowserTestCaseClients(clients);
      const browsers = flattenClientToBrowserVersion(execClients);
      expect(browsers).toMatchSnapshot();
    });

    afterAll(() => {
      process.env.BETA_BROWSERS = originalBetaBrowsers;
    });
  });

  // Mobile Browsers
  describe('getMobileTestCaseClients', () => {
    const clients = setBrowserStackMobileClients();
    it('should include a single iOS and Android device by default', () => {
      const execClients = getMobileTestCaseClients(clients);
      const browsers = flattenClientToDeviceName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it(`should augment additional devices when the 'DEFAULT' value is provided in options.versions`, () => {
      const execClients = getMobileTestCaseClients(clients, {
        versions: ['DEFAULT', 'ios 12'],
      });
      const browsers = flattenClientToDeviceName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it('should filter out platform and version based on skip options', () => {
      const execClients = getMobileTestCaseClients(clients, {
        skipPlatform: ['ios'],
        versions: ['DEFAULT', 'ios 12'],
      });
      const browsers = flattenClientToDeviceName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it('should filter out all platforms if wildcard is provided via skip options', () => {
      const execClients = getMobileTestCaseClients(clients, {
        skipPlatform: ['*'],
      });
      const browsers = flattenClientToDeviceName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it('should only target tablet devices', () => {
      const execClients = getMobileTestCaseClients(clients, {
        formFactors: ['tablet'],
      });
      const browsers = flattenClientToDeviceName(execClients);
      expect(browsers).toMatchSnapshot();
    });

    it('should only target samsung devices', () => {
      const execClients = getMobileTestCaseClients(clients, {
        skipPlatform: ['ios'],
        keyboards: ['samsung'],
      });
      const browsers = flattenClientToDeviceName(execClients);
      expect(browsers).toMatchSnapshot();
    });
  });
});
