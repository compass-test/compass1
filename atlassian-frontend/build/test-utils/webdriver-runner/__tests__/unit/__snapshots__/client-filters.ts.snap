// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserStack getBetaBrowserTestCaseClients should include only beta desktop browsers 1`] = `
Array [
  "chrome beta",
]
`;

exports[`BrowserStack getBrowserTestCaseClients should filter out all browsers when wildcard is passed via test case options 1`] = `Array []`;

exports[`BrowserStack getBrowserTestCaseClients should filter out browser(s) when skipped via test case options 1`] = `
Array [
  "firefox",
  "safari",
  "edge",
]
`;

exports[`BrowserStack getBrowserTestCaseClients should ignore invalid skipped browsers via test case options 1`] = `
Array [
  "chrome",
  "firefox",
  "safari",
  "edge",
]
`;

exports[`BrowserStack getBrowserTestCaseClients should include only desktop browsers 1`] = `
Array [
  "chrome",
  "firefox",
  "safari",
  "edge",
]
`;

exports[`BrowserStack getMobileTestCaseClients should augment additional devices when the 'DEFAULT' value is provided in options.versions 1`] = `
Array [
  "iOS 14 iPhone 12",
  "iOS 12 iPhone XS",
  "Android 10.0 Google Pixel 4",
]
`;

exports[`BrowserStack getMobileTestCaseClients should filter out all platforms if wildcard is provided via skip options 1`] = `Array []`;

exports[`BrowserStack getMobileTestCaseClients should filter out platform and version based on skip options 1`] = `
Array [
  "Android 10.0 Google Pixel 4",
]
`;

exports[`BrowserStack getMobileTestCaseClients should include a single iOS and Android device by default 1`] = `
Array [
  "iOS 14 iPhone 12",
  "Android 10.0 Google Pixel 4",
]
`;

exports[`BrowserStack getMobileTestCaseClients should only target samsung devices 1`] = `
Array [
  "Android 10.0 Samsung Galaxy S20",
]
`;

exports[`BrowserStack getMobileTestCaseClients should only target tablet devices 1`] = `
Array [
  "iOS 14 iPad Pro 12.9 2020",
]
`;
